pkg/test/framework/components/echo/native/native.go

```diff
@@ -0,0 +1,335 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package native
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/url"
+	"strconv"
+	"testing"
+
+	"github.com/hashicorp/go-multierror"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/test/application"
+	appEcho "istio.io/istio/pkg/test/application/echo"
+	"istio.io/istio/pkg/test/application/echo/proto"
+	"istio.io/istio/pkg/test/envoy"
+	"istio.io/istio/pkg/test/framework/components/echo"
+	"istio.io/istio/pkg/test/framework/components/echo/common"
+	"istio.io/istio/pkg/test/framework/components/environment/native"
+	"istio.io/istio/pkg/test/framework/components/pilot"
+	"istio.io/istio/pkg/test/framework/resource"
+)
+
+const (
+	envoyLogLovel = envoy.LogLevelWarning
+)
+
+var (
+	_ echo.Instance = &instance{}
+	_ io.Closer     = &instance{}
+)
+
+type instance struct {
+	id        resource.ID
+	endpoints []echo.Endpoint
+	client    *appEcho.Client
+	config    echo.Config
+
+	discoveryFilter discoveryFilter
+	app             application.Application
+	envoyMgr        envoyManager
+}
+
+// New creates a new native echo instance.
+func New(ctx resource.Context, cfg echo.Config) (out echo.Instance, err error) {
+	env := ctx.Environment().(*native.Environment)
+
+	// Fill in defaults for any missing values.
+	if err = common.FillInDefaults(ctx, &cfg); err != nil {
+		return nil, err
+	}
+
+	c := &instance{
+		config: cfg,
+	}
+	c.id = ctx.TrackResource(c)
+
+	// Convert the configured ports for the echo application. Ignore any specified port numbers.
+	appPorts := make(model.PortList, 0, len(cfg.Ports))
+	for _, p := range cfg.Ports {
+		appPorts = append(appPorts, &model.Port{
+			Name:     p.Name,
+			Protocol: p.Protocol,
+		})
+	}
+	appFactory := &appEcho.Factory{
+		Ports:   appPorts,
+		Version: c.config.Version,
+	}
+
+	if cfg.NoSidecar {
+		// No sidecar is simple - just create the app...
+
+		c.app, err = appFactory.NewApplication(application.DefaultDialer)
+		if err != nil {
+			return nil, err
+		}
+
+		// Update the configuration with the ports assigned by the application.
+		cfg.Ports = cfg.Ports[:0]
+		for _, p := range c.app.GetPorts() {
+			cfg.Ports = append(cfg.Ports, echo.Port{
+				Name:         p.Name,
+				Protocol:     p.Protocol,
+				ServicePort:  p.Port,
+				InstancePort: p.Port,
+			})
+		}
+	} else {
+		// Using a sidecar Envoy proxy. Need to wire up a custom discovery filter and start Envoy...
+
+		if cfg.Galley == nil {
+			return nil, fmt.Errorf("galley must be provided when running echo with a sidecar")
+		}
+		if cfg.Pilot == nil {
+			return nil, fmt.Errorf("pilot must be provided when running echo with a sidecar")
+		}
+
+		// Get Pilot's discovery address.
+		pilotDiscoveryAddress := cfg.Pilot.(pilot.Native).GetDiscoveryAddress()
+
+		// Create the discovery filter that modifies the XDS from Pilot to allow the application run
```

Comment:
nit: it is a little confusing to use filter in this context, just because of envoy filters, but I also can't think of a better word here